wrappers/dbus_test.go

```diff
@@ -0,0 +1,179 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package wrappers_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/testutil"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+type dbusTestSuite struct {
+	tempdir string
+}
+
+var _ = Suite(&dbusTestSuite{})
+
+func (s *dbusTestSuite) SetUpTest(c *C) {
+	s.tempdir = c.MkDir()
+	dirs.SetRootDir(s.tempdir)
+}
+
+func (s *dbusTestSuite) TearDownTest(c *C) {
+	dirs.SetRootDir("")
+}
+
+const dbusSnapYaml = `
+name: snapname
+version: 1.0
+slots:
+  system1:
+    interface: dbus
+    bus: system
+    name: org.example.Foo
+  system2:
+    interface: dbus
+    bus: system
+    name: org.example.Bar
+  session1:
+    interface: dbus
+    bus: session
+    name: org.example.Foo
+  session2:
+    interface: dbus
+    bus: session
+    name: org.example.Bar
+apps:
+  system-svc:
+    command: bin/start-system
+    daemon: simple
+    activates-on:
+      - system1
+      - system2
+  session-svc:
+    command: bin/start-session
+    daemon: simple
+    daemon-scope: user
+    activates-on:
+      - session1
+      - session2
+`
+
+func (s *dbusTestSuite) TestGenerateDBusActivationFile(c *C) {
+	info := snaptest.MockSnap(c, dbusSnapYaml, &snap.SideInfo{Revision: snap.R(12)})
+
+	app := info.Apps["system-svc"]
+	svcWrapper, err := wrappers.GenerateDBusActivationFile(app, "org.example.Foo")
+	c.Assert(err, IsNil)
+	c.Check(string(svcWrapper), Equals, `[D-BUS Service]
+Name=org.example.Foo
+Comment=Bus name for snap application snapname.system-svc
+SystemdService=snap.snapname.system-svc.service
+Exec=/usr/bin/snap run snapname.system-svc
+AssumedAppArmorLabel=snap.snapname.system-svc
+User=root
+X-Snap=snapname
+`)
+
+	app = info.Apps["session-svc"]
+	svcWrapper, err = wrappers.GenerateDBusActivationFile(app, "org.example.Foo")
+	c.Assert(err, IsNil)
+	c.Check(string(svcWrapper), Equals, `[D-BUS Service]
+Name=org.example.Foo
+Comment=Bus name for snap application snapname.session-svc
+SystemdService=snap.snapname.session-svc.service
+Exec=/usr/bin/snap run snapname.session-svc
+AssumedAppArmorLabel=snap.snapname.session-svc
+X-Snap=snapname
+`)
```

Comment:
@alexmurray asked if I wanted to look at this any more. I didn't look at this very closely and trust @alexmurray's review, but one thing did come to mind: I don't see in the PR negative tests for crafted values for these fields in the activation file. I think we should be ok since there are checks elsewhere that for example, make sure that the dbus name is valid, but having a check or two with invalid values for each field we are writing out would be a good idea to ensure all the dots stay connected going forward (the exhaustive checks can of course remain in the parsing code).

Reply:
For what it is worth, the data we're substituting in here has all been validated as part of loading the `snap.yaml`.

If the bus name in the slot definition does not pass `ValidateDBusBusName()`, then the dbus interface's `BeforePrepareSlot` method will fail.  This causes the slot to be marked as bad and deleted from the snap.  That in turn means we will not generate an activation file wrapper for the slot.

Reply:
I've added a unit test to demonstrate how invalid dbus slots will be ignored.